runtime/compare.c

```diff
@@ -99,6 +101,22 @@ static intnat compare_val(value v1, value v2, int total)
   return res;
 }
 
+static void poll_pending_actions(struct compare_stack* stk)
+{
+  if (caml_check_pending_actions()) {
+    value exn;
+    Begin_roots_block((value*)(stk->stack),
+                      (stk->limit - stk->stack) * sizeof(struct compare_stack) / sizeof(value));
```

Comment:
Are all these values valid ocaml values? Isn't it just from `stk->stack` to `sp`?

Reply:
Regarding your bug: even after fixing it, I still see some `0` roots in the roots block on the failing tests.

Reply:
Actually, I wonder if there may be an existing bug/glitch/weirdness in the indexing -- or maybe this is just the lack of sleep on my side hallucinating bugs, apologies. We would naively expect one of these invariants to hold, but which one?

1. the slots from `stk->stack` to `sp` *excluded* contain valid stack slots
2. the slots from `stk->stack` to `sp` *included* contain valid stack slots

My reading of the code is that we expect `sp` to point to the "current" stack slot, and we increment `sp` before filling a new slot, so "included" interpretation (2), but in fact we initialize `sp` to `stk->stack` without pushing an element at the beginning, so "excluded" interpretation (1) at initialization.

So my understanding is that in fact, `stk->stack` is allocated memory that plays the role of a sentinel value, and is never actually filled with any valid slack slot. We only ever dereference `sp` after incrementing it (so we are working on the non-zero slot) or after checking the sp value against the sentinel value, so this is correct, but we are missing one slot of our 8 initial slots, so this may limit performance slightly by causing reallocation earlier than necessary.

I need to fix the root range to stop at `sp` (thanks for this remark @smuenzel that set me on the right track), but for now I also need to have it actually start at `sp->stack + 1`.

If my analysis above is correct, I would consider removing the sentinel slot to get a simpler structure and obtain an extra slot.

Reply:
(see other threads)